### ISSUE TITLE
Component trait

### ISSUE THREAD

**Issue by cryptoquick** [hecrj/iced#648]
I see you have a Program trait, but this doesn't implement a "new" method:
https://github.com/hecrj/iced/blob/1f7e8b7f3d1804c39c8e0934b25f3ef178de269c/native/src/program.rs

First of all, what do you call a trait with `new`, `update`, and `view` methods? I don't see a quick answer in the TEA docs other than "program" or "module", is it true they don't really have an idiomatic new for this, like, say, a React Component that implements a `render` method and some event listeners that set local state on the component (either explicitly through calling `setState`, or implicitly through internally-managed state through Hooks)?

It doesn't seem to be appropriate to call this a "Widget", since this seems lower-level, and I don't think it's the responsibility of Widgets to manage state themselves, they're more for specific view components, or as React devs call this, a "stateless component". Stateless components don't call setState, and they technically shouldn't have React hooks called, and are generally used in simpler UI component libraries, similar to using a CSS framework but not the associated JS functionality.

For example, in my application, I want a convenience trait so I can call update methods on any Component that implements one. Ideally, there'd also be a default implementation that matches nothing, like a NoOp, while allowing update to be called with the provided Message generic argument. Such a trait would likely need another generic provided for the State, contained in the struct the Component trait implements.

Additionally, by standardizing this Component interface, I'm able to call view methods in my Application view function without needing to specify the individual implementations themselves.

Understandably, both probably needs to be called selectively, since I've also implemented a sort of Page routing logic. I prefer not to pollute my native application with terms drawing from web technologies, so I call these Pages rather than Routes or whatever you want to call the thing that cosmetically updates browser window history (though, routing might be helpful with a browser content, like a program compiled with a wasm target, but wasm has other uses beyond frontend now, of which Iced should also support for HTML static renderer deployment to things like CloudFlare Workers, or Lambda, or Krustlet; essentially full stack servers that return HTML strings after being provided a routing context). Regardless, I'm referring to the design idiom where you go see different pages or views after clicking or touching a nav bar, or on mobile, they have these bars typically on the bottom of the screen in iOS (and on the top in Android) that never really goes away as the user interacts with the app, which I call a Page Selector in my app.

TL;DR: What should a "component" be called, as per the definition provided above, and also, does such a concept already exist, or should it, or should these things be considered something that an application developer should implement, even though most of us will probably wind up reinventing the wheel individually in our own apps?

Oh, and one more thing. I'm sure there are things you'd prefer to not have added to every iced app, since it really would expand the scope of the UI framework to more of an application framework, which is understandably something a single maintainer would want to avoid supporting, maybe, but as you might know, I have much greater ambitions for Iced as a sort of unified application framework that supports a variety of targets and application contexts beyond what's already implemented. The goal being, to put it succinctly, the capability to implement the user interface for an entire Rust-native application in a _single codebase_ that compiles to a SPA, that also supports SSR, is easiiy deployed on FaaS cloud infra, and also, _within the same codebase_, with **no necessity for splitting the project up**, will support everything from terminal UI (TUI) apps, to native mobile devices compiled in aarch64 targets and using mobile-native UI frameworks. This is, essentially, the "holy grail" of application development, and Iced could be the "killer app" for Rust.

**Comment by twitchyliquid64** [comment]
I don't think there is an equivalent to react components - possibly intentionally. This has been discussed in the iced chatroom: [direct link](https://iced.zulipchat.com/#narrow/stream/213316-general/topic/Async.20component.20updates/near/216384098).

Direct quote of the relevant section for posterity:

> **Héctor Ramón**: The Elm Architecture has no notion of components. In fact, [thinking in terms of components is discouraged](https://guide.elm-lang.org/webapps/structure.html#components).

**Comment by cryptoquick** [comment]
This actually makes a lot of sense, and explains a lot about what I've been finding lately. Especially given Hector's explanation. I guess I need to think of other ways to solve my page routing / navigation feature, and comb through the examples more, maybe there's something there that can help me shape my thinking. Thanks for chiming in, @twitchyliquid64

**Comment by cryptoquick** [comment]
I had been hammering at this off and on all day and it's not as easy as it would first seem.

`component.rs`
```rust
use iced::{Command, Element};

pub trait Component: Sized {
  type Message;

  fn new() -> Self;

  fn update(&mut self, message: Self::Message) -> Command<Self::Message>;

  fn view(&mut self) -> Element<'_, Self::Message>;
}
```

I kept running into issues around the fact that I can't assign all that to a struct, because the "size can't be known at compile time", which kinda makes sense.

The problem I kept running into is, I'm not sure where I would have been able to assign my components, reference their view and update methods in useful ways, while also being able to return elements from the view method without referencing data owned only by the view method. And then there's the problem that I still wasn't able to get "Components" based on differing types and implementations to be considered functionally equivalent (like an interface) to other components, so I can use them interchangeably.

And then I realized... Splitting all this up into different files is not only premature, it's actively holding my application development back. All the work I'd done so far on my Iced app, carefully architecting it into different files, much as you would a JS and React project... These instincts led to work that wasn't really going to be useful, much less practical.

Thanks for the nudge in the right direction, @twitchyliquid64 :smile_cat:

**Comment by alex13sh** [comment]
Where is the subscription method? 
if I want to subscribe to a timer?

**Comment by cryptoquick** [comment]
@alex13sh Are you making fun of me

**Comment by cryptoquick** [comment]
Nevermind, I just realized there's a method I haven't seen in the examples...

https://docs.rs/iced/0.2.0/iced/trait.Application.html#method.subscription

Looks really useful, thank you. But, please, for the love of all that's good, don't do what I just spent all of yesterday on... :joy_cat:
